pkg/networkservice/common/dial/client.go

```diff
@@ -0,0 +1,111 @@
+// clienturlctx.ClientURL(ctx) and put the resulting grpc.ClientConnInterface into the ctx using clientconn.Store(..)
+// where it can be retrieved by other chain elements using clientconn.Load(...)
+package dial
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/postpone"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clientconn"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+)
+
+type dialClient struct {
+	chainCtx    context.Context
+	dialOptions []grpc.DialOption
+	dialTimeout time.Duration
+}
+
+// NewClient - returns new dial chain element
+func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
+	o := &option{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &dialClient{
+		chainCtx:    chainCtx,
+		dialOptions: o.dialOptions,
+		dialTimeout: o.dialTimeout,
+	}
+}
+
+func (d *dialClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	closeContextFunc := postpone.ContextWithValues(ctx)
+	// If no clientURL, we have no work to do
+	// call the next in the chain
+	clientURL := clienturlctx.ClientURL(ctx)
+	if clientURL == nil {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+
+	cc, _ := clientconn.LoadOrStore(ctx, newDialer(d.chainCtx, d.dialTimeout, d.dialOptions...))
+
+	// If there's an existing grpc.ClientConnInterface and it's not ours, call the next in the chain
+	di, ok := cc.(*dialer)
+	if !ok {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+
+	// If our existing dialer has a different URL close down the chain
+	if di.clientURL != nil && di.clientURL.String() != clientURL.String() {
+		closeCtx, closeCancel := closeContextFunc()
+		defer closeCancel()
+		_ = di.Dial(closeCtx, di.clientURL)
+		_, _ = next.Client(ctx).Close(clienturlctx.WithClientURL(closeCtx, di.clientURL), request.GetConnection(), opts...)
+	}
+
+	err := di.Dial(ctx, clientURL)
+	if err != nil {
+		clientconn.Delete(ctx)
+		closeCtx, closeCancel := closeContextFunc()
+		defer closeCancel()
+		_, _ = next.Client(ctx).Close(clienturlctx.WithClientURL(closeCtx, di.clientURL), request.GetConnection(), opts...)
+		return nil, err
+	}
+
+	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		_ = di.Close()
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (d *dialClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*emptypb.Empty, error) {
+	cc, _ := clientconn.LoadOrStore(ctx, newDialer(ctx, d.dialTimeout, d.dialOptions...))
+
+	di, ok := cc.(*dialer)
+	if !ok {
+		return next.Client(ctx).Close(ctx, conn, opts...)
+	}
+	defer func() {
```

Comment:
Probably we additionally need to make `di.Dial()` here?

Reply:
Yes... fixed here: https://github.com/networkservicemesh/sdk/pull/1083/commits/af236c83ecf8d096f62934ea64f1b4aac9ffed03